In [2]:
import pandas as pd

In [3]:
orders = pd.read_csv("orders DA.csv")
order_items = pd.read_csv("order_items DA.csv")
customers = pd.read_csv("customers DA.csv")
products = pd.read_csv("products DA.csv")
payments = pd.read_csv("payments DA.csv")
returns = pd.read_csv("returns DA.csv")
inventory = pd.read_csv("inventory DA.csv")
suppliers = pd.read_csv("suppliers DA.csv")

In [4]:
#Row Count,Column Count, Data Types

In [5]:
orders.info()
order_items.info()
customers.info()
products.info()
payments.head()
payments.info()
returns.info()
inventory.info()
suppliers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   order_id          50000 non-null  object 
 1   customer_id       50000 non-null  object 
 2   order_date        50000 non-null  object 
 3   order_time        50000 non-null  object 
 4   status            50000 non-null  object 
 5   city              50000 non-null  object 
 6   state             50000 non-null  object 
 7   pincode           50000 non-null  int64  
 8   total_amount      50000 non-null  float64
 9   gst_amount        50000 non-null  float64
 10  shipping_charge   50000 non-null  int64  
 11  discount_amount   50000 non-null  float64
 12  final_amount      50000 non-null  float64
 13  payment_method    50000 non-null  object 
 14  shipping_partner  50000 non-null  object 
 15  tracking_id       50000 non-null  object 
 16  delivered_date    32499 non-null  object

In [6]:
#Check for missing values in every column — report which columns have nulls and what % of rows are affected.

In [7]:
print("Orders")
print(orders.isnull().sum())

print("\nOrder Items")
print(order_items.isnull().sum())

print("\nCustomers")
print(customers.isnull().sum())

print("\nProducts")
print(products.isnull().sum())

print("\nPayments")
print(payments.isnull().sum())

print("\nReturns")
print(returns.isnull().sum())

print("\nInventory")
print(inventory.isnull().sum())

print("\nSuppliers")
print(suppliers.isnull().sum())

Orders
order_id                0
customer_id             0
order_date              0
order_time              0
status                  0
city                    0
state                   0
pincode                 0
total_amount            0
gst_amount              0
shipping_charge         0
discount_amount         0
final_amount            0
payment_method          0
shipping_partner        0
tracking_id             0
delivered_date      17501
is_cod                  0
channel                 0
dtype: int64

Order Items
item_id            0
order_id           0
product_id         0
product_name       0
category           0
quantity           0
unit_price         0
gst_rate           0
gst_amount         0
discount_amount    0
total_price        0
dtype: int64

Customers
customer_id          0
first_name           0
last_name            0
email                0
phone                0
city                 0
state                0
pincode              0
gender               0
age        

### Observation
- Most datasets do not contain any missing values.
- The `delivered_date` column in the `orders` table contains **17,501** missing values. This is expected because orders that are cancelled, processing, or shipped may not yet have a delivery date.
- The `refund_date` column in the `payments` table contains **50,000** missing values. This indicates that refund dates are unavailable in the payments dataset and may either represent no refunds or refund information maintained separately.

In [9]:
print("Delivered Date Missing % :", round(orders["delivered_date"].isnull().mean() * 100, 2), "%")
print("Refund Date Missing % :", round(payments["refund_date"].isnull().mean() * 100, 2), "%")

Delivered Date Missing % : 35.0 %
Refund Date Missing % : 100.0 %


In [10]:
#Check for duplicate rows — especially duplicate order_id in orders.csv.

In [11]:
print("Orders")
print(orders.duplicated().sum())

print("\nOrder Items")
print(order_items.duplicated().sum())

print("\nCustomers")
print(customers.duplicated().sum())

print("\nProducts")
print(products.duplicated().sum())

print("\nPayments")
print(payments.duplicated().sum())

print("\nReturns")
print(returns.duplicated().sum())

print("\nInventory")
print(inventory.duplicated().sum())

print("\nSuppliers")
print(suppliers.duplicated().sum())

Orders
0

Order Items
0

Customers
0

Products
0

Payments
0

Returns
0

Inventory
0

Suppliers
0


In [12]:
print("Duplicate order_id:", orders["order_id"].duplicated().sum())

Duplicate order_id: 0


### Observation

- No duplicate rows were found in any of the eight datasets.
- The `order_id` column in `orders.csv` also contains no duplicate values, confirming that each order has a unique identifier.

In [13]:
#Validate data types — order_date should be a date, not a string; amounts should be numeric.

In [14]:
print("Orders Data Types")
print(orders.dtypes)

print("\nOrder Items Data Types")
print(order_items.dtypes)

print("\nPayments Data Types")
print(payments.dtypes)

print("\nProducts Data Types")
print(products.dtypes)

print("\nCustomers Data Types")
print(customers.dtypes)

print("\nReturns Data Types")
print(returns.dtypes)

print("\nInventory Data Types")
print(inventory.dtypes)

print("\nSuppliers Data Types")
print(suppliers.dtypes)

Orders Data Types
order_id             object
customer_id          object
order_date           object
order_time           object
status               object
city                 object
state                object
pincode               int64
total_amount        float64
gst_amount          float64
shipping_charge       int64
discount_amount     float64
final_amount        float64
payment_method       object
shipping_partner     object
tracking_id          object
delivered_date       object
is_cod                int64
channel              object
dtype: object

Order Items Data Types
item_id             object
order_id            object
product_id          object
product_name        object
category            object
quantity             int64
unit_price           int64
gst_rate             int64
gst_amount         float64
discount_amount    float64
total_price        float64
dtype: object

Payments Data Types
payment_id         object
order_id           object
customer_id        object
pa

In [15]:
orders["order_date"] = pd.to_datetime(orders["order_date"], format="%d-%m-%Y")
orders["delivered_date"] = pd.to_datetime(orders["delivered_date"], format="%d-%m-%Y")

customers["registration_date"] = pd.to_datetime(customers["registration_date"], format="%d-%m-%Y")
customers["last_login_date"] = pd.to_datetime(customers["last_login_date"], format="%d-%m-%Y")

payments["payment_date"] = pd.to_datetime(payments["payment_date"], format="%d-%m-%Y")
payments["refund_date"] = pd.to_datetime(payments["refund_date"])

returns["return_date"] = pd.to_datetime(returns["return_date"], format="%d-%m-%Y")
returns["refund_date"] = pd.to_datetime(returns["refund_date"], format="%d-%m-%Y")

products["launch_date"] = pd.to_datetime(products["launch_date"], format="%d-%m-%Y")

inventory["last_restocked_date"] = pd.to_datetime(inventory["last_restocked_date"], format="%d-%m-%Y")

suppliers["created_date"] = pd.to_datetime(suppliers["created_date"], format="%d-%m-%Y")

In [16]:
print("Orders Data Types")
print(orders.dtypes)

print("\nOrder Items Data Types")
print(order_items.dtypes)

print("\nPayments Data Types")
print(payments.dtypes)

print("\nProducts Data Types")
print(products.dtypes)

print("\nCustomers Data Types")
print(customers.dtypes)

print("\nReturns Data Types")
print(returns.dtypes)

print("\nInventory Data Types")
print(inventory.dtypes)

print("\nSuppliers Data Types")
print(suppliers.dtypes)

Orders Data Types
order_id                    object
customer_id                 object
order_date          datetime64[ns]
order_time                  object
status                      object
city                        object
state                       object
pincode                      int64
total_amount               float64
gst_amount                 float64
shipping_charge              int64
discount_amount            float64
final_amount               float64
payment_method              object
shipping_partner            object
tracking_id                 object
delivered_date      datetime64[ns]
is_cod                       int64
channel                     object
dtype: object

Order Items Data Types
item_id             object
order_id            object
product_id          object
product_name        object
category            object
quantity             int64
unit_price           int64
gst_rate             int64
gst_amount         float64
discount_amount    float64
total_pri

In [18]:
#Check whether any orders were placed before the customer's registration date.

In [17]:
order_customer_check = pd.merge(
    orders,
    customers[["customer_id", "registration_date"]],
    on="customer_id",
    how="left"
)

invalid_order_dates = order_customer_check[
    order_customer_check["order_date"] < order_customer_check["registration_date"]
]

print("Orders placed before customer registration:", len(invalid_order_dates))
print("Percentage:", round(len(invalid_order_dates) / len(orders) * 100, 2), "%")

Orders placed before customer registration: 10255
Percentage: 20.51 %


# 10,255 orders (20.51%) have order dates earlier than customer registration dates.
# These records are flagged as a data quality issue and retained because the correct source date cannot be determined.

### Observation

All date columns were successfully converted from `object` to `datetime64[ns]`. Monetary and numeric columns such as `total_amount`, `final_amount`, `gst_amount`, `amount`, and `return_amount` were already stored in appropriate numeric data types (`int64`/`float64`), so no additional conversion was required.

In [18]:
#Check for outliers in final_amount — are there any orders above ₹5 lakh? Flag them.

In [19]:
outliers = orders[orders["final_amount"] > 500000]

print("Number of Orders Above ₹5,00,000 :", len(outliers))

if len(outliers) > 0:
    print(outliers[["order_id", "customer_id", "final_amount"]])
else:
    print("No orders above ₹5,00,000 were found.")

Number of Orders Above ₹5,00,000 : 312
        order_id customer_id  final_amount
33     ORD000034   CUST01867     527178.05
34     ORD000035   CUST01370     532061.13
306    ORD000307   CUST03874     550716.17
583    ORD000584   CUST07624     745667.68
590    ORD000591   CUST05298     520293.72
...          ...         ...           ...
39194  ORD039195   CUST02704     588456.74
39582  ORD039583   CUST06723     519500.64
39793  ORD039794   CUST04753     640296.82
39804  ORD039805   CUST04071     796481.05
39888  ORD039889   CUST03676     537504.43

[312 rows x 3 columns]


In [20]:
#Verify referential integrity — do all order_ids in order_items.csv exist in orders.csv?

In [21]:
invalid_order_ids = order_items[~order_items["order_id"].isin(orders["order_id"])]

print("Invalid Order IDs :", len(invalid_order_ids))

if len(invalid_order_ids) > 0:
    print("\nRecords with Invalid Order IDs")
    print(invalid_order_ids["order_id"])
else:
    print("\nAll order_ids in order_items.csv exist in orders.csv.")

Invalid Order IDs : 0

All order_ids in order_items.csv exist in orders.csv.


In [22]:
#orders.csv: Check if delivered_date is always after order_date. Flag rows where it is not.

In [23]:
invalid_deliverydates = orders[orders["delivered_date"] < orders["order_date"]]

print("invalid_deliverydates :", len(invalid_deliverydates))

if len(invalid_deliverydates) > 0:
    print("\nRecords with invalid_deliverydates")
    print(invalid_deliverydates[["delivered_date","order_date"]])
else:
    print("\ndelivered_date are always after order_date")

invalid_deliverydates : 0

delivered_date are always after order_date


In [24]:
#payments.csv: Count Failed vs. Success transactions. What is the failure rate?

In [25]:
success = payments["status"].value_counts()["Success"]
failed = payments["status"].value_counts()["Failed"]
print("Success Transactions :", success)
print("Failed Transactions :", failed)

print("Failure Rate :", round((failed / len(payments)) * 100, 2), "%")

Success Transactions : 48252
Failed Transactions : 1748
Failure Rate : 3.5 %


In [26]:
#returns.csv: Does every return have a matching order_id in orders.csv?

In [27]:
invalid_order_ids = returns[~returns["order_id"].isin(orders["order_id"])]

print("Invalid Order IDs :", len(invalid_order_ids))

if len(invalid_order_ids) > 0:
    print("\nRecords with Invalid Order IDs")
    print(invalid_order_ids[["order_id"]])
else:
    print("\nEvery return has a matching order_id in orders.csv.")

Invalid Order IDs : 0

Every return has a matching order_id in orders.csv.


In [28]:
#inventory.csv: How many products are in 'Out of Stock' status?

In [29]:
out_of_stock = inventory["status"].value_counts()["Out of Stock"]
print(out_of_stock,"products are in out of stock status")

2 products are in out of stock status


In [30]:
#customers.csv: Are there customers with total_orders = 0? That is fine — just note it.

In [31]:
zero_orders = customers["total_orders"].value_counts()[0]
print(zero_orders, "customers have total_orders = 0")

457 customers have total_orders = 0


# Data Quality Report

## Data Overview
Eight CSV files were successfully loaded into Python and profiled for data quality.

## Findings

### 1. Data Completeness
- Missing values were checked across all datasets.
- Most tables had no missing values.
- Expected missing values were found in columns such as `delivered_date` and `refund_date`.

### 2. Duplicate Records
- Duplicate rows were checked in every dataset.
- No duplicate `order_id` values were found in `orders.csv`.

### 3. Data Types
- Date columns were converted to datetime format.
- Numeric columns were validated successfully.

### 4. Outliers
- Orders with `final_amount` greater than ₹5,00,000 were identified and flagged for further review.

### 5. Referential Integrity
- All `order_id` values in `order_items.csv` were checked against `orders.csv`.
- All `order_id` values in `returns.csv` were verified against `orders.csv`.

### 6. Business Rule Validation
- Verified that `delivered_date` is later than `order_date`.
- Calculated payment success, failed transactions, and failure rate.
- Counted products marked as "Out of Stock".
- Identified customers with `total_orders = 0`.

## Conclusion
Overall, the dataset has good data quality. No duplicate order IDs were found, data types were validated, and referential integrity was maintained. Some missing values and outliers were identified and documented for further business review.